File: ClassCodes of c++ implementations/recursion.cpp
#include <iostream>
using namespace std;

bool isSorted(int input[], int n) {
    // Special/Edge case
    if(n == 0) {
        return true;
    }
    // Base case
    if(n == 1) {
        return true;
    }

    // Small calculation
    if(input[0] > input[1]) {
        return false;
    }

    // Recursive call
    bool smallAns = isSorted(input + 1, n - 1);
    return smallAns;
}

bool isSorted_2(int *input, int n) {
    if(n <= 1) {
        return true;
    }
    if(input[n-2] > input[n-1]) {
        return false;
    }

    bool smallAns = isSorted_2(input, n-1);
    return smallAns;
}

// start = 1
void print(int n, int start = 1) {
    if(start == n + 1) {
        return;
    }
    cout << start << " ":
    print(n, start + 1);
}


void print_1(int n) {
    if(n == 0) {
        return;
    }
    print(n - 1);
    cout << n << " ";
}


















int main() {
    
}

File: class codes of java implementations/L8/Recursion.java
package L8;

public class Recursion {

	public static void funA(int x) {
		x += 10;
	}
	
	public static int factorial(int n) {
		// Base case
		if(n <= 1) {
			return 1;
		}
		// Recursive call
		int x = factorial(n - 1);
		
		// Small calculation
		int ans = n * x;
		
		return ans;
	}
	
	public static void print(int n) {
		if(n <= 0) {
			return;
		}
		
		print(n - 1);
		
		System.out.println(n);
	}
	
	public static int fibonacci(int n) {
		// Base case
		if(n == 0) {
			return 0;
		}
		if(n == 1) {
			return 1;
		}
		
		// Recursive call
		int a = fibonacci(n - 1);
		int b = fibonacci(n - 2);
		
		// Small calculation
		return a + b;
	}
	
	public static void main(String[] args) {
		print(5);
//		System.out.println(factorial(4));
		/*int x = 10;
		funA(x);
		System.out.println(x);*/
	}
}
